mm/kfence/static_branch.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/atomic.h>
+#include <linux/mm.h>
+#include <linux/static_key.h>
+#include <linux/wait.h>
+#include <linux/workqueue.h>
+
+#include "kfence.h"
+#include "../slab.h"
+
+/* The static key to set up a KFENCE allocation. */
+DEFINE_STATIC_KEY_FALSE(kfence_allocation_key);
+
+/* Gates the allocation, ensuring only one succeeds in a given period. */
+static atomic_t allocation_gate = ATOMIC_INIT(1);
+/* Wait queue to wake up heartbeat timer task. */
+static DECLARE_WAIT_QUEUE_HEAD(allocation_wait);
+
+void *kfence_alloc_with_size(struct kmem_cache *s, size_t size, gfp_t flags)
+{
+	void *ret;
+
+	if (atomic_inc_return(&allocation_gate) > 1)
```

Comment:
I assume we can have the static key enabled for some milliseconds (schedule the delayed work, send IPIs, poke text, etc). During all that time all allocations on all CPUs will hammer this atomic var.
I think it can make sense to read and check it first. The single sampled allocation doing an additional read is not noticeable, but potentially thousands of subsequent allocs not doing the contented increment can make some difference.
```suggestion
	if (atomic_read(&allocation_gate) || atomic_inc_return(&allocation_gate) > 1)
```

Reply:
Very good point, done.